### ISSUE TITLE
Add to repo previously-existing files on git-lfs server

### ISSUE THREAD

**Issue by ivotron** [git-lfs/git-lfs#1148]
Is it possible to point `git-lfs` to an existing server and add existing files to a repo?

**Comment by technoweenie** [comment]
The Git LFS command line tool only talks to servers implementing the LFS API as described [here](https://github.com/github/git-lfs/tree/master/docs/api). Does that answer your question? If not, can you be more specific?

**Comment by ivotron** [comment]
Sorry for the bad wording. What I mean is if it is possible to have a list of files already in a git-lfs server, and create a git repo from an arbitrary subset of them. So instead of creating a repo and then ingesting files in it, I wonder if it makes sense to go the other way. I hope it's clear.

**Comment by technoweenie** [comment]
A git server can reference whatever LFS blobs it wants. It really depends how the LFS server is implemented, though. On GitHub, objects in different repository networks are not shared.

**Comment by ivotron** [comment]
By "git server" do you mean git repo? or git-lfs server? Please let me rephrase my question in more concrete terms. This is an example of what the content of a lfs-tracked file looks like:

```
version https://git-lfs.github.com/spec/v1
oid sha256:7d5fdacdd614cd076bc552c2ee395438303c4ed1b03723d85e6f7e01db5ae5e6
size 120
```

If I manually create references like this in a repo, will git-lfs behave as if I added them by doing `git add`? Also, looking at the API spec, looks there is no way of listing the objects for an endpoint, am I right? Would it be possible to know this in some other way?

**Comment by technoweenie** [comment]
I meant "git repo". But since every clone is a full repo, I was _technically_ correct :trollface:.

> If I manually create files like this in a repo, and I create a corresponding .gitconfig (or .lfsconfig file), will git-lfs behave as if I added them by doing git add? The goal would be to add files to a repo that already exist in a git-lfs server. Or is the assumption that one file is owned by a single repo? I can see that it might be problematic to share Oids between multiple repos since the server would need to have a reference counter.

This is also _technically_ possible. Our docs don't really specify what the LFS server has to do behind the scenes. Though you bring up great points, and it'd be nice to have suggestions for implementers.

I would say that any lfs server run by a git host (I'll use GitHub as an example because I know first hand) _must_ track which repositories have which objects. So no, what you're asking for is not possible on GitHub, and I hope not other git hosts that support lfs.

You, as part of some private group, can definitely implement a custom lfs server that allowed your group of repositories to share objects. You'd set it's url in `.lfsconfig`, and every collaborator would be on the same page. You won't hurt my feelings if you host your git repo on GitHub, but store your files somewhere that works better for your team :)

> Also, looking at the API spec, looks there is no way of listing the objects in an lfs-server, am I right? Would it be possible to know this in some other way?

There isn't, but there could be. We've been focusing on what we _need_ for normal lfs operations, and working the kinks out there. But I do see room for other commonly useful api endpoints, such as listing or deleting objects. Keep in mind it'll take some time for any added endpoints to be adopted by the various implementations...

**Comment by ttaylorr** [comment]
Hi @ivotron, I think that the above response should mostly answer what's going on here, in that you can push whatever Git LFS objects you want to any repository that you have push access to, but there is no pre-ordained mechanism for taking Git LFS objects from one repository and mirroring them in another.

If you have any questions in the future in this vein or otherwise, please do not hesitate to let me know about them here! 😄.
